### ISSUE TITLE
Notifications: services notify when host down

### ISSUE THREAD

**Issue by UrBnW** [centreon/centreon-engine#286]
Hi,

I'm facing this strange issue :

![68206820-f4af0480-ffcd-11e9-9fd7-749b5443ca64](https://user-images.githubusercontent.com/40244829/68380262-c822f600-014f-11ea-845e-5e05ad216e18.png)

As you can see, _sys-backup_ service notifies even if its host is down.
In comparison, _sys-uptime_ service does not notify.
Here, _sys-backup_ should then not have notified.

Perhaps a bug somewhere ?

Edit : rootcause here below : https://github.com/centreon/centreon-engine/issues/286#issuecomment-568037963

Many thanks 👍

**Comment by SylvestreG** [comment]
I think the service has been notified because the host was DOWN in soft state.

If the host was DOWN in hard state, the service will not be notified.

**Comment by CPbN** [comment]
Thank you @SylvestreG.
I'm not sure about your answer : if service notifies when host is in soft down state, then `sys-uptime` should have notified, but it did not 😕

**Comment by CPbN** [comment]
Look, again this morning, same behavior :

![Screen Shot 2019-11-06 at 17 10 08](https://user-images.githubusercontent.com/40244829/68315833-acb8dc00-00b8-11ea-8465-1a48407427a5.png)

Services do not notify while host is (soft) down, but `sys-backup`, in this example.
Rather strange.

**Comment by CPbN** [comment]
Here are 4 OK situations where the service does not notify when host is (soft / hard) down.

![OK1](https://user-images.githubusercontent.com/40244829/68380068-6f535d80-014f-11ea-8c14-bde361c93b3c.png)
![OK2](https://user-images.githubusercontent.com/40244829/68380070-6f535d80-014f-11ea-9a72-8a1edb44e029.png)
![OK3](https://user-images.githubusercontent.com/40244829/68380071-6febf400-014f-11ea-92ef-09a659c4218f.png)
![OK4](https://user-images.githubusercontent.com/40244829/68380072-6febf400-014f-11ea-8fe8-c4a2447cf719.png)

**Comment by CPbN** [comment]
And 4 KO situations, where the service notifies when the host is down.

![KO1](https://user-images.githubusercontent.com/40244829/68380132-8bef9580-014f-11ea-838a-ad19b3e53181.png)
![KO2](https://user-images.githubusercontent.com/40244829/68380134-8bef9580-014f-11ea-8497-01e63e1de6db.png)
![KO3](https://user-images.githubusercontent.com/40244829/68380135-8bef9580-014f-11ea-8108-f862295f7d7f.png)
![KO4](https://user-images.githubusercontent.com/40244829/68380136-8c882c00-014f-11ea-8e10-5213a47054ca.png)

**Comment by CPbN** [comment]
These are 8 different hosts, inheriting from the same template, so identically configured.
Rather strange then that this behavior looks "random"...

**Comment by bouda1** [comment]
Hi @CPbN,

In the Engine configuration, click on the Debug tab and in the Debug Level
field, activate the notification information. Then set the Debug Verbosity
to "More detailed information".
Save this configuration and restart engine.

In /var/log/centreon-engine/centengine.debug you should get informations
telling why notifications are sent or not. It should help us to understand
your problem.
Thanks.

**Comment by CPbN** [comment]
OK, issue reproduced, here it is :

![Screen Shot 2019-11-18 at 22 39 39](https://user-images.githubusercontent.com/40244829/69098999-04acf680-0a5a-11ea-90cb-6d94d5dc0b86.png)

And related log lines :

```
[1574096096] [25958] Centreon Engine 19.10.6 starting ... (PID=25958)
[1574096096] [25958] Local time is Mon Nov 18 17:54:56 2019
[1574096096] [25958] LOG VERSION: 2.0
[1574096096] [25958] Calculating next valid notification time...
[1574096096] [25958] Default interval: 0
[1574096096] [25958] Interval used for calculating next valid notification time: 0
[1574096096] [25958] Event broker module '/usr/lib64/centreon-engine/externalcmd.so' initialized successfully
[1574096096] [25958] Centreon Broker: log applier: applying 1 logging objects
[1574096096] [25958] Event broker module '/usr/lib64/nagios/cbmod.so' initialized successfully
[1574096096] [25958] INITIAL HOST STATE: fw01;UP;HARD;1;OK - fw01 rta 12.929ms lost 0%
[1574096096] [25958] INITIAL SERVICE STATE: fw01;sys-backup;WARNING;HARD;3;WARNING: Backup age: 437248.9h
[1574112271] [25958] SERVICE ALERT: fw01;sys-backup;UNKNOWN;HARD;3;UNKNOWN: SNMP Session : unable to create
[1574112271] [25958] Processed notification command: /bin/sh -c '...'
[1574112271] [25958] SERVICE NOTIFICATION: Admin;fw01;sys-backup;UNKNOWN;service-notify-by-email;UNKNOWN: SNMP Session : unable to create
[1574112271] [25958] Processed notification command: /bin/sh -c '...'
[1574112271] [25958] SERVICE NOTIFICATION: Tools;fw01;sys-backup;UNKNOWN;service-notify-by-email;UNKNOWN: SNMP Session : unable to create
[1574112444] [25958] SERVICE ALERT: fw01;sys-backup;WARNING;HARD;3;WARNING: Backup age: 437253.5h
```

**Comment by lpinsivy** [comment]
@CPbN this is not the correct debug file. It should be centengine.debug

Regards

**Comment by CPbN** [comment]
@lpinsivy it is `centengine.debug`, in `/var/log/centreon-engine/`.
I just cross-checked again.
Thank you.

**Comment by CPbN** [comment]
@lpinsivy it is `centengine.debug`, in `/var/log/centreon-engine/`.
I just cross-checked again.
Of course feel free if you need mode log / info etc...
For sure this issue is rather weird 😕
Thank you.

**Comment by CPbN** [comment]
OK, I've found the root cause.
The concerned service for which we receive unexpected `UNKNOWN` alerts is already in a `HARD` `!OK` state before switching to `UNKNOWN`. This is the key.

It's easily reproductible :
Let's assume a service in `HARD` `WARNING` state.
You've already received a `WARNING` notification for this.
The related host becomes `!UP`.
The previous service then switches from `WARNING` to `UNKNOWN` state.
And here you'll receive an unexpected `UNKNOWN` notification.

Could this then be corrected please ?

Thank you 👍

**Comment by lpinsivy** [comment]
Hi @CPbN 

I'm not sure that checking any Notification filters (checkboxes) for sys-uptime = all option.
For me is look like NONE

**Comment by CPbN** [comment]
Hi @lpinsivy,
It's all, as per the documentation, and this service correctly notifies.
I was, at time of the first post, not sure about the root cause.
However, I have found it, as per my post just below : https://github.com/centreon/centreon-engine/issues/286#issuecomment-566149922
You should then manage to easily reproduce this issue.
Thank you again !

**Comment by CPbN** [comment]
OK, I've found the root cause.
The concerned service for which we receive unexpected `UNKNOWN` alerts is already in a `HARD` state before switching to `UNKNOWN`. This is the key.
Well, to be correct, the key is that these services then have `tries == max_check_attempts`.

It's easily reproductible :
Let's assume a service in `HARD` `WARNING` state (`tries == max_check_attempts`).
You've already received a `WARNING` notification for this.
The related host becomes `!UP` (tested with `DOWN`, not with `UNREACHABLE`, should be the same).
The previous service then switches from `WARNING` to `UNKNOWN` state.
And here, if host is `SOFT` `!UP`, you'll receive an unexpected `UNKNOWN` notification (notification is not sent if host is already in `HARD` state).

I say unexpected, because services which have `tries != max_check_attempts` (for example generally `OK` services) won't get their `tries` incremented when host is `SOFT` `!UP`. So these services won't never notify about their state change.
Whereas, then, a service which already has `tries == max_check_attempts` will.

This issue can also be reproduced with 2 `OK` services, first one with `max_check_attempts == 1`, second one with `max_check_attempts == 2`.
First one will notify when host is `SOFT` `!UP`, second will never (whatever number of retries done).

There may then be an issue in which order checks are done in engine before notification is sent.

Could this then be corrected please ?

Thank you 👍

**Comment by UrBnW** [comment]
Hi @bouda1, any progress on this annoying bug ?
Many thanks 👍

**Comment by UrBnW** [comment]
Still in 19.10.18.

**Comment by bouda1** [comment]
Hi @UrBnW We are going to work on it very soon!! :100:

**Comment by omercier** [comment]
Hi @UrBnW,
Have you tried setting `Soft Service Dependencies Option` to "yes"?
![image](https://user-images.githubusercontent.com/32134301/125455253-006f3d26-63bc-41f3-91e1-c32c67671876.png)
Here is what it is supposed to do:
`This option determines whether or not Monitoring Engine will use soft state information when checking host and service dependencies. Normally Monitoring Engine will only use the latest hard host or service state when checking dependencies. If you want it to use the latest state (regardless of whether its a soft or hard state type), enable this option.`
In my understanding it should fit your need, do you confirm?

**Comment by omercier** [comment]
Ok, thank you for your answer.
The notification should not be sent with a SOFT DOWN host if this parameter is set to "yes", but they are. 
The issue is in our backlog and should be fixed in the next weeks or months.

**Comment by omercier** [comment]
Hi Benjamin,
Yes, this issue has been fixed for centreon-engine 20.10.7 and 21.04.4 (and of course in 21.10.0).
Ref MON-10642

**Comment by UrBnW** [comment]
Hi Olivier,
Pretty good news, many thanks 👍
